Framework/Algorithms/src/ReflectometryReductionOne2.cpp

```diff
@@ -460,6 +464,60 @@ void ReflectometryReductionOne2::exec() {
 }
 
 /**
+* Utility function to create a unique workspace name for diagnostic outputs
+* based on a given input workspace name
+*
+* @param inputName [in] : the input name
+* @return : the output name
+*/
+std::string ReflectometryReductionOne2::createDebugWorkspaceName(
+    const std::string &inputName) {
+  std::string result = inputName;
+
+  if (summingInQ()) {
+    if (getPropertyValue("ReductionType") == "DivergentBeam")
+      result += "_QSD"; // Q-summed divergent beam
+    else
+      result += "_QSB"; // Q-summed bent (non-flat) sample
+  } else {
+    result += "_LS"; // lambda-summed
+  }
+
+  return result;
+}
+
+/**
+* Utility function to add the given workspace to the ADS if debugging is
+* enabled.
+*
+* @param ws [in] : the workspace to add to the ADS
+* @param wsName [in] : the name to output the workspace as
+* @param wsSuffix [in] : a suffix to apply to the wsName
+* @param debug [in] : true if the workspace should be added to the ADS (the
+* function does nothing if this is false)
+* @param step [inout] : the current step number, which is added to the wsSuffix
+* and is incremented after the workspace is output
+*/
+void ReflectometryReductionOne2::outputDebugWorkspace(
+    MatrixWorkspace_sptr ws, const std::string &wsName,
+    const std::string &wsSuffix, const bool debug, int &step) {
+  // Nothing to do if debug is not enabled
+  if (debug) {
+    // Clone the workspace because otherwise we can end up outputting the same
+    // workspace twice with different names, which is confusing.
+    auto cloneAlg = this->createChildAlgorithm("CloneWorkspace");
+    cloneAlg->initialize();
+    cloneAlg->setProperty("InputWorkspace", ws);
+    cloneAlg->execute();
+    auto cloneWS = cloneAlg->getProperty("OutputWorkspace");
```

Comment:
Thanks for the tip, I've made that change (although without `auto` because I need a `shared_ptr` rather than a `unique_ptr`).